File: packages/imanikurd-quran/src/tafsir.ts
import { loadData } from "./loadData.js";
import type { TafsirEntry, TafsirOption } from "./types.js";

/** All 13 Kurdish Tafsir options available in the package */
export const TAFSIR_OPTIONS: TafsirOption[] = [
  { id: "rebar", name: "تەفسیری ڕێبەر", file: "tafsir_rebar.json" },
  { id: "asan", name: "تەفسیری ئاسان", file: "tafsir_asan.json" },
  { id: "puxta", name: "تەفسیری پوختە", file: "tafsir_puxta.json" },
  { id: "raman", name: "تەفسیری ڕامان", file: "tafsir_raman.json" },
  { id: "hazhar", name: "تەفسیری هەژار (مامۆستا هەژار)", file: "tafsir_hazhar.json" },
  { id: "zhin", name: "تەفسیری ژیان", file: "tafsir_zhin.json" },
  { id: "sanahi", name: "تەفسیری سەناهی", file: "tafsir_sanahi.json" },
  { id: "maisar", name: "تەفسیری مویەسەر", file: "tafsir_maisar.json" },
  { id: "tawhid", name: "تەفسیری تەوحیدی", file: "tafsir_tawhid.json" },
  { id: "roshn", name: "تەفسیری ڕۆشن", file: "tafsir_roshn.json" },
  { id: "runahi", name: "تەفسیری ڕوناهی", file: "tafsir_runahi.json" },
  { id: "mokhtasar", name: "تەفسیری موختەسەر", file: "tafsir_mokhtasar.json" },
  { id: "krd", name: "تەفسیری کوردی", file: "tafsir_krd.json" },
];

const tafsirCache = new Map<string, TafsirEntry[]>();

function normalizeEntry(item: { s: number | string; a: number; t: string }): TafsirEntry {
  return {
    s: typeof item.s === "string" ? parseInt(item.s, 10) : item.s,
    a: item.a,
    t: item.t,
  };
}

function getTafsirFile(tafsirId: string): string {
  const option = TAFSIR_OPTIONS.find((t) => t.id === tafsirId);
  if (!option) {
    throw new Error(
      `Unknown tafsir id "${tafsirId}". Valid ids: ${TAFSIR_OPTIONS.map((t) => t.id).join(", ")}`
    );
  }
  return option.file;
}

function loadTafsirRaw(tafsirId: string): TafsirEntry[] {
  if (tafsirCache.has(tafsirId)) {
    return tafsirCache.get(tafsirId)!;
  }

  const file = getTafsirFile(tafsirId);
  const raw = loadData<Array<{ s: number | string; a: number; t: string }>>(file);
  const entries = raw.map(normalizeEntry);
  tafsirCache.set(tafsirId, entries);
  return entries;
}

/** Get list of available Tafsir options */
export function getTafsirOptions(): TafsirOption[] {
  return TAFSIR_OPTIONS;
}

/** Get full tafsir dataset for a given tafsir id */
export function getTafsir(tafsirId: string): TafsirEntry[] {
  return loadTafsirRaw(tafsirId);
}

/** Get tafsir text for a specific ayah */
export function getTafsirForAyah(
  tafsirId: string,
  surahNumber: number,
  ayahNumber: number
): string | undefined {
  const entries = loadTafsirRaw(tafsirId);
  const entry = entries.find((e) => e.s === surahNumber && e.a === ayahNumber);
  return entry?.t;
}

/** Get-get all tafsir entries for a surah as a map of ayah number -> text */
export function getTafsirForSurah(
  tafsirId: string,
  surahNumber: number
): Record<number, string> {
  const entries = loadTafsirRaw(tafsirId);
  const result: Record<number, string> = {};

  for (const entry of entries) {
    if (entry.s === surahNumber) {
      result[entry.a] = entry.t;
    }
  }

  return result;
}

/** Search tafsir text across all entries of a given tafsir */
export function searchTafsir(tafsirId: string, query: string): TafsirEntry[] {
  const normalized = query.trim();
  if (!normalized) return [];

  return loadTafsirRaw(tafsirId).filter((e) => e.t.includes(normalized));
}

export type { TafsirEntry, TafsirOption };

File: packages/imanikurd_flutter/lib/src/tafsir.dart
import 'load_data.dart';
import 'models.dart';

const List<TafsirOption> tafsirOptions = [
  TafsirOption(id: 'rebar', name: 'تەفسیری ڕێبەر', file: 'tafsir_rebar.json'),
  TafsirOption(id: 'asan', name: 'تەفسیری ئاسان', file: 'tafsir_asan.json'),
  TafsirOption(id: 'puxta', name: 'تەفسیری پوختە', file: 'tafsir_puxta.json'),
  TafsirOption(id: 'raman', name: 'تەفسیری ڕامان', file: 'tafsir_raman.json'),
  TafsirOption(id: 'hazhar', name: 'تەفسیری هەژار (مامۆستا هەژار)', file: 'tafsir_hazhar.json'),
  TafsirOption(id: 'zhin', name: 'تەفسیری ژیان', file: 'tafsir_zhin.json'),
  TafsirOption(id: 'sanahi', name: 'تەفسیری سەناهی', file: 'tafsir_sanahi.json'),
  TafsirOption(id: 'maisar', name: 'تەفسیری مویەسەر', file: 'tafsir_maisar.json'),
  TafsirOption(id: 'tawhid', name: 'تەفسیری تەوحیدی', file: 'tafsir_tawhid.json'),
  TafsirOption(id: 'roshn', name: 'تەفسیری ڕۆشن', file: 'tafsir_roshn.json'),
  TafsirOption(id: 'runahi', name: 'تەفسیری ڕوناهی', file: 'tafsir_runahi.json'),
  TafsirOption(id: 'mokhtasar', name: 'تەفسیری موختەسەر', file: 'tafsir_mokhtasar.json'),
  TafsirOption(id: 'krd', name: 'تەفسیری کوردی', file: 'tafsir_krd.json'),
];

final Map<String, List<TafsirEntry>> _tafsirCache = {};

TafsirOption _getTafsirOption(String tafsirId) {
  for (final option in tafsirOptions) {
    if (option.id == tafsirId) return option;
  }
  throw ArgumentError(
    'Unknown tafsir id "$tafsirId". Valid ids: ${tafsirOptions.map((t) => t.id).join(", ")}',
  );
}

Future<List<TafsirEntry>> _loadTafsirRaw(String tafsirId) async {
  if (_tafsirCache.containsKey(tafsirId)) {
    return _tafsirCache[tafsirId]!;
  }

  final file = _getTafsirOption(tafsirId).file;
  final raw = await loadDataAsync<List<dynamic>>(file);
  final entries = raw.map((e) => TafsirEntry.fromJson(e as Map<String, dynamic>)).toList();
  _tafsirCache[tafsirId] = entries;
  return entries;
}

List<TafsirOption> getTafsirOptions() => tafsirOptions;

Future<List<TafsirEntry>> getTafsir(String tafsirId) => _loadTafsirRaw(tafsirId);

Future<String?> getTafsirForAyah(String tafsirId, int surahNumber, int ayahNumber) async {
  final entries = await _loadTafsirRaw(tafsirId);
  for (final entry in entries) {
    if (entry.s == surahNumber && entry.a == ayahNumber) return entry.t;
  }
  return null;
}

Future<Map<int, String>> getTafsirForSurah(String tafsirId, int surahNumber) async {
  final entries = await _loadTafsirRaw(tafsirId);
  final result = <int, String>{};
  for (final entry in entries) {
    if (entry.s == surahNumber) {
      result[entry.a] = entry.t;
    }
  }
  return result;
}

Future<List<TafsirEntry>> searchTafsir(String tafsirId, String query) async {
  final normalized = query.trim();
  if (normalized.isEmpty) return [];
  final entries = await _loadTafsirRaw(tafsirId);
  return entries.where((e) => e.t.contains(normalized)).toList();
}
